### PR TITLE
*: add external flag for testing external networks

### LINKED ISSUE #160: Add ability to test external networks
Right now we always run Docker images locally, but we also need some ability to test networks that are already running. Keys can be synchronized between loader and network.

### PR DESCRIPTION
**Pull request by AliceInHunterland**
`./runner.sh -e -d "Go4x1" -m rate -q 1000 -z 5m -t 30s -a 192.168.1.100:20331 -a 192.168.1.101:20331`

Close #160

### CODE REVIEW

File: runner.sh [hunk 1]
```diff
@@ -263,8 +268,31 @@ if [ -n "$NEOBENCH_VOTE" ]; then
...
+      export CGO_ENABLED=0
+      export GO111MODULE=on
+
+      go build -v -o neo-bench -trimpath
+
+      popd
+
+      ./cmd/bench/neo-bench -o "$LOG_OUTPUT" "${ARGS[@]}"> "./bench_${OUTPUT}_${MODE}_${COUNT}.log" 2>&1
+
+#  docker compose -f "${DC_BENCH}" run --rm base-bench neo-bench -o "$LOG_OUTPUT" "${ARGS[@]}" > "./bench_${OUTPUT}_${MODE}_${COUNT}.log" 2>&1
```
Comment (AliceInHunterland) [inline runner.sh:290]: delete

File: README.md [hunk 2]
```diff
@@ -194,6 +194,9 @@ $ ./runner.sh --nodes mixed -d "MixedGoRPC4x1" -m rate -q 50 -z 5m -t 30s
                                    Example: -t 30s --request_timeout 15s (default 30s)
   -i, --in                         Path to input file to load transactions.
                                    Example: -i ./dump.txs --in /path/to/import/transactions
+      --vote                       Vote before the bench.
+      --disable-stats              Disable stats output.
+
```
Comment (AliceInHunterland) [inline README.md:199]: extra line

File: Makefile [hunk 3]
```diff
@@ -118,6 +118,7 @@ config:
 # Generate transactions, dump and nodes configurations for four-nodes network
 prepare: stop config $(BUILD_DIR)/dump.$(NEOBENCH_TYPE).$(NEOBENCH_FROM_COUNT).$(NEOBENCH_TO_COUNT).txs
 
+prepare-local: config $(BUILD_DIR)/dump.$(NEOBENCH_TYPE).$(NEOBENCH_FROM_COUNT).$(NEOBENCH_TO_COUNT).txs
```
Comment (AnnaShaleva) [inline Makefile:121]: Not needed, `make stop` won't hurt anyway.

File: README.md [hunk 4]
```diff
@@ -194,6 +194,9 @@ $ ./runner.sh --nodes mixed -d "MixedGoRPC4x1" -m rate -q 50 -z 5m -t 30s
                                    Example: -t 30s --request_timeout 15s (default 30s)
   -i, --in                         Path to input file to load transactions.
                                    Example: -i ./dump.txs --in /path/to/import/transactions
+      --vote                       Vote before the bench.
+      --disable-stats              Disable stats output.
```
Comment (AnnaShaleva) [inline README.md:198]: `--disable-stats` deserves a separate commit.

Comment (AnnaShaleva) [inline README.md:198]: > Disable stats output.

Which stats? What is stats? "Disable memory and CPU usage statistics collection." or something like that.

File: README.md [hunk 5]
```diff
@@ -194,6 +194,9 @@ $ ./runner.sh --nodes mixed -d "MixedGoRPC4x1" -m rate -q 50 -z 5m -t 30s
                                    Example: -t 30s --request_timeout 15s (default 30s)
   -i, --in                         Path to input file to load transactions.
                                    Example: -i ./dump.txs --in /path/to/import/transactions
+      --vote                       Vote before the bench.
```
Comment (AnnaShaleva) [inline README.md:197]: Not related to this commit.

File: runner.sh [hunk 6]
```diff
@@ -3,13 +3,14 @@
 source .env
 
 OUTPUT=""
-ARGS=(-i "/dump.txs")
+ARGS=()
```
Comment (AnnaShaleva) [inline runner.sh:6]: Why?

File: runner.sh [hunk 7]
```diff
@@ -252,7 +257,7 @@ else
 	fatal "Invalid validator count: $NEOBENCH_VALIDATOR_COUNT"
 fi
 
-OUTPUT="/out/${OUTPUT}_${MODE}_${COUNT}.log"
+LOG_OUTPUT="/out/${OUTPUT}_${MODE}_${COUNT}.log"
```
Comment (AnnaShaleva) [inline runner.sh:260]: Why?

File: runner.sh [hunk 8]
```diff
@@ -263,8 +268,31 @@ if [ -n "$NEOBENCH_VOTE" ]; then
 	ARGS+=(--vote)
 fi
 
-make prepare
 
-docker compose "${FILES[@]}" run bench neo-bench -o "$OUTPUT" "${ARGS[@]}"
+if [ "$EXTERNAL_NETWORK" = true ]; then
```
Comment (AnnaShaleva) [inline runner.sh:272]: `if [ "$EXTERNAL_NETWORK" ]; then`?

File: runner.sh [hunk 9]
```diff
@@ -263,8 +268,31 @@ if [ -n "$NEOBENCH_VOTE" ]; then
...
+
+#  docker compose -f "${DC_BENCH}" run --rm base-bench neo-bench -o "$LOG_OUTPUT" "${ARGS[@]}" > "./bench_${OUTPUT}_${MODE}_${COUNT}.log" 2>&1
+else
+    ARGS+=(-i "/dump.txs")
+    make prepare
+
+    docker compose "${FILES[@]}" run bench neo-bench -o "$LOG_OUTPUT" "${ARGS[@]}"
+
+    make stop
+fi
```
Comment (AnnaShaleva) [inline runner.sh:298]: No new line in the end of the file.

File: runner.sh [hunk 10]
```diff
@@ -263,8 +268,31 @@ if [ -n "$NEOBENCH_VOTE" ]; then
...
+
+      go build -v -o neo-bench -trimpath
+
+      popd
+
+      ./cmd/bench/neo-bench -o "$LOG_OUTPUT" "${ARGS[@]}"> "./bench_${OUTPUT}_${MODE}_${COUNT}.log" 2>&1
+
+#  docker compose -f "${DC_BENCH}" run --rm base-bench neo-bench -o "$LOG_OUTPUT" "${ARGS[@]}" > "./bench_${OUTPUT}_${MODE}_${COUNT}.log" 2>&1
+else
+    ARGS+=(-i "/dump.txs")
```
Comment (AnnaShaleva) [inline runner.sh:292]: What if it was overriden by user before?

File: runner.sh [hunk 11]
```diff
@@ -263,8 +268,31 @@ if [ -n "$NEOBENCH_VOTE" ]; then
...
-make stop
+      make prepare-local
+
+      pushd "$(dirname "$0")/cmd/bench" || exit
+
+      export GOGC=off
+      export CGO_ENABLED=0
+      export GO111MODULE=on
+
+      go build -v -o neo-bench -trimpath
```
Comment (AnnaShaleva) [inline runner.sh:284]: Need a separate target to build bench locally.

Comment (AnnaShaleva) [reply runner.sh:284]: Since you don't need to rebuild it every time when script is called.

File: runner.sh (hunk 10 above)
Comment (AliceInHunterland) [reply runner.sh:292]: The runner.sh script does not have the -i flag, so the user cannot change it without modifying the code.

File: runner.sh (hunk 7 above)
Comment (AliceInHunterland) [reply runner.sh:260]: because it's easier to save bench logs to the file to prevent losing it.
`./neo-bench -o "$LOG_OUTPUT" "${ARGS[@]}"> "./bench_${OUTPUT}_${MODE}_${COUNT}.log" 2>&1`

File: Makefile [hunk 12]
```diff
@@ -39,6 +39,15 @@ help:
 # Build all images
 build: gen build.node.bench build.node.go build.node.sharp
 
+# Build Benchmark binary file
+binary-bench:
+	@cd /cmd/bench && \
+		GOGC=off \
+		CGO_ENABLED=0 \
+		GO111MODULE=on \
+		go build -v -o neo-bench -trimpath
```
Comment (AnnaShaleva) [inline Makefile:48]: Make the output be `cmd/bin/bench`

Comment (AnnaShaleva) [reply Makefile:48]: Add this `bin` to .gitignore.

File: Makefile [hunk 13]
```diff
@@ -39,6 +39,15 @@ help:
 # Build all images
 build: gen build.node.bench build.node.go build.node.sharp
 
+# Build Benchmark binary file
+binary-bench:
+	@cd /cmd/bench && \
```
Comment (AnnaShaleva) [inline Makefile:44]: Don't perform `cd`, just add the target argument to `go build`.

File: Makefile [hunk 14]
```diff
@@ -39,6 +39,15 @@ help:
...
+		GOGC=off \
+		CGO_ENABLED=0 \
+		GO111MODULE=on \
```
Comment (AnnaShaleva) [inline Makefile:47]: These vars should be exported. Like `export GOGC=off`.

Comment (AnnaShaleva) [inline Makefile:47]: Why do you need `GO111MODULE`? It's outdated and unused anymore.

Comment (AnnaShaleva) [reply Makefile:47]: So it should be:
```
	@set -x \
		&& export GOGC=off \
		&& export CGO_ENABLED=0 \
...
```

File: Makefile (hunk 12 above)
Comment (AnnaShaleva) [reply Makefile:48]: Add `-trimpath` argument?

File: Makefile [hunk 15]
```diff
@@ -39,6 +39,15 @@ help:
 # Build all images
 build: gen build.node.bench build.node.go build.node.sharp
 
+# Build Benchmark binary file
+binary-bench:
```
Comment (AnnaShaleva) [inline Makefile:43]: Rename it to `build.bench`.

Comment (AnnaShaleva) [inline Makefile:43]: Create a separate commit with this new make target.

File: runner.sh (hunk 8 above)
Comment (AnnaShaleva) [reply runner.sh:281]: Does it work?

File: runner.sh [hunk 16]
```diff
@@ -273,7 +278,18 @@ if [ -n "$NEOBENCH_VOTE" ]; then
 fi
 
 make prepare
```
Comment (AnnaShaleva) [inline runner.sh:280]: Remove `make prepare` then.

File: runner.sh (hunk 8 above)
Comment (AliceInHunterland) [reply runner.sh:281]: no

File: runner.sh (hunk 6 above)
Comment (AliceInHunterland) [reply runner.sh:6]: moved it to the 281 line, because '/dump.txs' binds to docker, for the local run without docker we need the local path

File: .gitignore [hunk 17]
```diff
@@ -11,3 +11,4 @@ vendor
 .docker/rpc/go.protocol*.yml
 .docker/rpc/sharp.*.json
 .docker/build/*.acc
+cmd/bin/
```
Comment (AnnaShaleva) [inline .gitignore:14]: No new line in the end of the file.

File: Makefile [hunk 18]
```diff
@@ -39,6 +39,14 @@ help:
 # Build all images
 build: gen build.node.bench build.node.go build.node.sharp
 
+# Build Benchmark binary file
+build.binary.bench:
```
Comment (AnnaShaleva) [inline Makefile:43]: Why not `build.bench`?

File: Makefile [hunk 19]
```diff
@@ -39,6 +39,14 @@ help:
...
+	@set -x \
+		&& export GOGC=off \
+		&& export CGO_ENABLED=0 \
+		&& cd cmd \
```
Comment (AnnaShaleva) [inline Makefile:47]: Remove `cd`, we don't need it. Use `-C`, read the https://pkg.go.dev/cmd/go.

File: cmd/bench/main.go [hunk 20]
```diff
@@ -43,6 +43,7 @@ func main() {
 		timeLimit       = v.GetDuration("timeLimit")
 		mode            = internal.BenchMode(v.GetString("mode"))
 		client          *internal.RPCClient
+		disableStats    = v.GetBool("disable-stats")
```
Comment (AnnaShaleva) [inline cmd/bench/main.go:46]: Move it to a separate commit, external mode and nostats are two separate logical changes.

File: runner.sh [hunk 21]
```diff
@@ -273,7 +278,17 @@ if [ -n "$NEOBENCH_VOTE" ]; then
...
-make stop
+      ./cmd/bin/bench -o "$OUTPUT" "${ARGS[@]}"
+
+else
+    ARGS+=(-i "/dump.txs")
+
+    docker compose "${FILES[@]}" run bench neo-bench -o "$OUTPUT" "${ARGS[@]}"
+
+    make stop
+fi
```
Comment (AnnaShaleva) [inline runner.sh:294]: A lot of unnecessary empty new lines.

File: runner.sh [hunk 22]
```diff
@@ -273,7 +278,17 @@ if [ -n "$NEOBENCH_VOTE" ]; then
...
 
 make prepare
+if [ "$EXTERNAL_NETWORK" = true ]; then
+      ARGS+=(-i "./.docker/build/dump.$NEOBENCH_TYPE.$NEOBENCH_FROM_COUNT.$NEOBENCH_TO_COUNT.txs" --disable-stats)
 
-docker compose "${FILES[@]}" run bench neo-bench -o "$OUTPUT" "${ARGS[@]}"
+      make build.binary.bench
 
-make stop
+      ./cmd/bin/bench -o "$OUTPUT" "${ARGS[@]}"
```
Comment (AnnaShaleva) [inline runner.sh:286]: We need a graceful way to stop the process after the end. Did you always use `kill -9` manually?

File: runner.sh [hunk 23]
```diff
@@ -273,7 +278,17 @@ if [ -n "$NEOBENCH_VOTE" ]; then
 fi
 
 make prepare
+if [ "$EXTERNAL_NETWORK" = true ]; then
```
Comment (AnnaShaleva) [inline runner.sh:281]: Please, add more detailed documentation to the README about this mode.

File: runner.sh [hunk 24]
```diff
@@ -273,7 +278,17 @@ if [ -n "$NEOBENCH_VOTE" ]; then
...
+      ARGS+=(-i "./.docker/build/dump.$NEOBENCH_TYPE.$NEOBENCH_FROM_COUNT.$NEOBENCH_TO_COUNT.txs" --disable-stats)
 
-docker compose "${FILES[@]}" run bench neo-bench -o "$OUTPUT" "${ARGS[@]}"
+      make build.binary.bench
```
Comment (AnnaShaleva) [inline runner.sh:284]: And BTW, do you really need to build it here? runner.sh is not responsible for building. Include `build.binary.bench` into `build` target and adjust corresponding documentation in the README.

File: Makefile (hunk 15 above)
Comment (AnnaShaleva) [reply Makefile]: Not resolved.

Comment (AnnaShaleva) [reply Makefile]: Not resolved.

File: Makefile (hunk 13 above)
Comment (AnnaShaleva) [reply Makefile]: Not resolved.

File: README.md [hunk 25]
```diff
@@ -307,13 +307,15 @@ The following default configurations are available:
...
+   -e, --external                   Use external network for benchmarking. Default is false. -a flag should be used to specify RPC addresses.
 
 ```
 
 ## Build options
 
 By default, neo-bench uses released versions of Neo nodes to build Docker images.
-However, you can easily test non-released branches or even separate commits for both Go and C# Neo nodes.
+However, you can easily test non-released branches or even separate commits for both Go and C# Neo nodes. 
+Also with `--external` flag benchmark will be build and run locally without docker containers.
```
Comment (AnnaShaleva) [inline README.md:318]: `--external` flag should be used for external network benchmarks to run Bench in a standalone mode without Docker container.

File: runner.sh [hunk 26]
```diff
@@ -279,8 +284,22 @@ if [ -n "$NEOBENCH_VOTE" ]; then
...
-docker compose "${FILES[@]}" run bench neo-bench -o "$OUTPUT" "${ARGS[@]}"
+trap cleanup SIGINT SIGTERM EXIT
 
-make stop
+make prepare
+if [ "$EXTERNAL_NETWORK" = true ]; then
+      ARGS+=(-i "./.docker/build/dump.$NEOBENCH_TYPE.$NEOBENCH_FROM_COUNT.$NEOBENCH_TO_COUNT.txs" --disable-stats)
+      ./cmd/bin/bench -o "$OUTPUT" "${ARGS[@]}"&
+      pid=$!
+      wait $pid
```
Comment (AnnaShaleva) [inline runner.sh:300]: Does it work? Can we just add some code that uses lsof + grep to filter out the benchmark-related processes to `make stop`?

File: Makefile [hunk 27]
```diff
@@ -37,7 +37,15 @@ help:
...
-build: gen build.node.bench build.node.go build.node.sharp
+build: gen build.node.bench build.node.go build.node.sharp build.bench
+
+# Build Benchmark binary file
+build.bench:
+	@echo "=> Building Bench binary file"
+	@set -x \
+		&& export GOGC=off \
+		&& export CGO_ENABLED=0 \
+		&& go -C cmd build -v -o bin/bench -trimpath ./bench
```
Comment (AnnaShaleva) [inline Makefile:48]: `go build -C cmd ...`

Comment (AnnaShaleva) [reply Makefile:48]: And please, be more attentive, testing workflow is still failing. Update it to the newer Go version (not from go.mod) and create an issue to rollback these changes once we migrate to the newer go:
```
=> Building Bench binary file
+ export GOGC=off
+ GOGC=off
+ export CGO_ENABLED=0
+ CGO_ENABLED=0
+ go -C cmd build -v -o bin/bench -trimpath ./bench
flag provided but not defined: -C
Go is a tool for managing Go source code.
```

File: README.md [hunk 28]
```diff
@@ -176,6 +176,48 @@ $ ./runner.sh --validators 1 --nodes sharp -d "SharpSingle" -m rate -q 25 -z 5m
 $ ./runner.sh --nodes mixed -d "MixedGoRPC4x1" -m rate -q 50 -z 5m -t 30s
 ...
 ```
+## Usage example (local benchmark + external cluster)
+
+1. Build the benchmark images and binary file with the following command:
```
Comment (AnnaShaleva) [inline README.md:181]: A comment is needed before the first step:

NeoBench can be run in a stand-alone mode (loader only) to benchmark some external network. NeoBench expects external network to be launched with the known set of validators and committee (with wallets from `./docker/ir/`) and not contain any transactions in the network (to successfully perform initial NEO/GAS transfers). In this setup the loader will be launched as a system process, without Docker container. You have to provide RPS address(-es) of some node from the external network to the loader instance on start. The loader will use the provided RPC to send transactions to the network.

Comment (AnnaShaleva) [inline README.md:181]: 1. Build the benchmark binary file with the following command:

File: README.md [hunk 29]
```diff
@@ -176,6 +176,48 @@ $ ./runner.sh --validators 1 --nodes sharp -d "SharpSingle" -m rate -q 25 -z 5m
...
+creating: 879937303/go.protocol.template.yml
+creating: 879937303/go.protocol.template.yml
+creating: 879937303/sharp.protocol.template.yml
+creating: 879937303/sharp.protocol.template.yml
+```
+
+3. Run Neo nodes with generated configuration files. For example latest version of [neo-go](https://github.com/nspcc-dev/neo-go/releases) can be used:
+```
+$ ./neo-go node -p
+```
```
Comment (AnnaShaleva) [inline README.md:215]: You don't need this, it's not a part of the Bench.

File: README.md [hunk 30]
```diff
@@ -176,6 +176,48 @@ $ ./runner.sh --validators 1 --nodes sharp -d "SharpSingle" -m rate -q 25 -z 5m
...
+
+4. Run benchmarks using the `runner.sh` script with passed RPC addresses and `--external` flag set:
```
Comment (AnnaShaleva) [inline README.md:217]: with passed RPC addresses -> with RPC address(-es) of the external network

File: README.md [hunk 31]
```diff
@@ -307,13 +349,15 @@ The following default configurations are available:
...
                                     Example: --msPerBlock 3000
+   -e, --external                   Use external network for benchmarking. Default is false. -a flag should be used to specify RPC addresses.
 
 ```
 
 ## Build options
 
 By default, neo-bench uses released versions of Neo nodes to build Docker images.
 However, you can easily test non-released branches or even separate commits for both Go and C# Neo nodes.
+`--external` flag should be used for external network benchmarks to run bench in a standalone mode without Docker container.
```
Comment (AnnaShaleva) [inline README.md:360]: to run bench -> to run the loader

File: README.md [hunk 32]
```diff
@@ -176,6 +176,50 @@ $ ./runner.sh --validators 1 --nodes sharp -d "SharpSingle" -m rate -q 25 -z 5m
 $ ./runner.sh --nodes mixed -d "MixedGoRPC4x1" -m rate -q 50 -z 5m -t 30s
 ...
 ```
+## Usage example (local benchmark + external cluster)
+
+NeoBench can be run in a stand-alone mode (loader only) to benchmark some external network. NeoBench expects external network to be launched with the known set of validators and committee (with wallets from `./docker/ir/`) and not contain any transactions in the network (to successfully perform initial NEO/GAS transfers). In this setup the loader will be launched as a system process, without Docker container. You have to provide RPS address(-es) of some node from the external network to the loader instance on start. The loader will use the provided RPC to send transactions to the network.
```
Comment (AnnaShaleva) [inline README.md:181]: A single line? Nice.

File: README.md (hunk 29 above)
Comment (AnnaShaleva) [reply README.md:217]: Not resolved.

**Review by AnnaShaleva — APPROVED** [review]
